File: app/src/main/java/com/spartan/karanbir/gallery/MainActivity.java
package com.spartan.karanbir.gallery;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.spartan.karanbir.gallery.adapter.ImagePagerAdapter;
import com.spartan.karanbir.gallery.adapter.OnImageClickedListener;
import com.spartan.karanbir.gallery.adapter.VideoPagerAdapter;
import com.spartan.karanbir.gallery.model.Animal;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author karanbir
 * @since 5/20/17
 */

public class MainActivity extends AppCompatActivity implements ViewPager.OnPageChangeListener, OnImageClickedListener {
    private static final String TAG = MainActivity.class.getSimpleName();
    private VideoPagerAdapter mVideoPagerAdapter;
    private ImagePagerAdapter mImagePagerAdapter;
    private ArrayList<Animal> mAnimals;
    @BindView(R.id.video_pager)
    ViewPager mVideoPager;
    @BindView(R.id.image_pager)
    ViewPager mImagePager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        Retrofit retrofit = setupRetrofit(getString(R.string.base_url));
        callApi(retrofit);
        mAnimals = new ArrayList<>();
        mImagePager.setPageMargin(-250);
        mImagePager.setOffscreenPageLimit(3);
        mVideoPagerAdapter = new VideoPagerAdapter(this, mAnimals);
        mImagePagerAdapter = new ImagePagerAdapter(this, mAnimals);
        mImagePagerAdapter.setOnImageClickedListener(this);
        mImagePager.setAdapter(mImagePagerAdapter);
        mVideoPager.setAdapter(mVideoPagerAdapter);
        mImagePager.setPageTransformer(false, new ViewPager.PageTransformer() {
            @Override
            public void transformPage(View view, float position) {
                final float normalizedPosition = Math.abs(Math.abs(position) - 1);
                view.setScaleX(normalizedPosition / 2 + 0.5f);
                view.setScaleY(normalizedPosition / 2 + 0.5f);
            }
        });

        mImagePager.addOnPageChangeListener(this);
        mVideoPager.addOnPageChangeListener(this);
    }


    private Retrofit setupRetrofit(String baseUrl) {
        Gson gson = new GsonBuilder().create();
        return new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();

    }

    private void callApi(Retrofit retrofit) {
        AnimalService animalApi = retrofit.create(AnimalService.class);
        Call<List<Animal>> animalCall = animalApi.getAnimals();
        animalCall.enqueue(new Callback<List<Animal>>() {
            @Override
            public void onResponse(Call<List<Animal>> call, Response<List<Animal>> response) {
                int statusCode = response.code();
                Log.d(TAG, Integer.toString(statusCode));
                if (statusCode == 200) {
                    mAnimals.addAll(response.body());
                    mImagePagerAdapter.setData(mAnimals);
                    mVideoPagerAdapter.setData(mAnimals);
                }
            }

            @Override
            public void onFailure(Call<List<Animal>> call, Throwable t) {

            }
        });
    }

    @Override
    public void onPageScrolled(int i, float v, int i1) {

    }

    @Override
    public void onPageSelected(int i) {
        mImagePager.setCurrentItem(i, true);
        mVideoPager.setCurrentItem(i, true);
    }

    @Override
    public void onPageScrollStateChanged(int i) {

    }

    @Override
    public void onImageClicked(int position) {
        mImagePager.setCurrentItem(position, true);
    }
}

File: app/src/main/java/com/spartan/karanbir/gallery/MainActivity.kt
package com.spartan.karanbir.gallery

import android.os.Bundle
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.google.gson.GsonBuilder
import com.spartan.karanbir.gallery.adapter.ImagePagerAdapter
import com.spartan.karanbir.gallery.adapter.OnImageClickedListener
import com.spartan.karanbir.gallery.adapter.VideoPagerAdapter
import com.spartan.karanbir.gallery.model.Animal
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.*

class MainActivity : AppCompatActivity(), ViewPager.OnPageChangeListener, OnImageClickedListener {

    companion object {
        val TAG: String = MainActivity::class.java.canonicalName
    }

    private var mVideoPagerAdapter: VideoPagerAdapter? = null
    private var mImagePagerAdapter: ImagePagerAdapter? = null
    private var mAnimals: ArrayList<Animal>? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //Setting up Retrofit and call api
        callApi(retrofit = setupRetrofit(getString(R.string.base_url)))
        image_pager.pageMargin = -250
        image_pager.offscreenPageLimit = 3
        mAnimals = Collections.EMPTY_LIST as ArrayList<Animal>?
        mImagePagerAdapter = ImagePagerAdapter(this,mAnimals)
        mVideoPagerAdapter = VideoPagerAdapter(this,mAnimals)
        mImagePagerAdapter?.setOnImageClickedListener(this)
        image_pager.adapter = mImagePagerAdapter
        video_pager.adapter = mVideoPagerAdapter

        image_pager.setPageTransformer(false, ViewPager.PageTransformer { view, position ->
            val normalizedPosition = Math.abs(Math.abs(position) - 1)
            view.scaleX = normalizedPosition / 2 + 0.5f
            view.scaleY = normalizedPosition / 2 + 0.5f
        })

        image_pager.addOnPageChangeListener(this)
        image_pager.addOnPageChangeListener(this)
    }

    private fun setupRetrofit(baseUrl: String?): Retrofit {
        val gson = GsonBuilder().create()
        return Retrofit.Builder().
                baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
    }

    private fun callApi(retrofit: Retrofit?) {
        val animalApi = retrofit?.create(AnimalService::class.java)
        val animalCall = animalApi?.getAnimals()

        animalCall?.enqueue(object : Callback<List<Animal>> {
            override fun onResponse(call: Call<List<Animal>>, response: Response<List<Animal>>) {
                val statusCode = response.code()
                Log.d(TAG, Integer.toString(statusCode))
                if (statusCode == 200) {
                    mAnimals?.addAll(response.body() as List<Animal>)
                    if(mAnimals != null){
                        mImagePagerAdapter?.setData(mAnimals!!)
                        mVideoPagerAdapter?.setData(mAnimals!!)
                    }

                }
            }

            override fun onFailure(call: Call<List<Animal>>, t: Throwable) {

            }
        })
    }


    override fun onPageScrollStateChanged(state: Int) {
        TODO("not implemented")
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        TODO("not implemented")
    }

    override fun onPageSelected(position: Int) {
        image_pager.setCurrentItem(position, true)
        video_pager.setCurrentItem(position, true)
    }

    override fun onImageClicked(position: Int) {
        image_pager.setCurrentItem(position, true)
    }
}


